PDEP-14: Dedicated string data type for pandas 3.0 (#58551)
pandas-dev/pandas: web/pandas/pdeps/00xx-string-dtype.md

# PDEP-XX: Dedicated string data type for pandas 3.0

- Created: May 3, 2024
- Status: Under discussion
- Discussion:
- Author: [Joris Van den Bossche](https://github.com/jorisvandenbossche)
- Revision: 1

## Abstract

This PDEP proposes to introduce a dedicated string dtype that will be used by
default in pandas 3.0:

* In pandas 3.0, enable a "string" dtype by default, using PyArrow if available

>   or otherwise the numpy object-dtype alternative.

Reviewer: Should you allow the possability of a NumPy 2 improved type for pandas 3?  With a heirarchy arrow -> np 2 -> np object?

Author: This proposal does not preclude any further improvements for the numpy-based string dtype using numpy 2.0. A few lines below I explicitly mention it as a future improvement and in the *"Object-dtype "fallback" implementation"* section as well.

I just don't want to explicitly commit to anything for pandas 3.0 related to that, given it is hard to judge right now how well it will work / how much work it is to get it ready. If it is ready by 3.0, then we can evaluate that separately, but this proposal doesn't stand or fall with it.

Regardless of whether to also use numpy 2.0, we have to agree on 1) making a "string" dtype the default for 3.0, 2) the missing value behaviour to use for this dtype, and 3) whether to provide an alternative for PyArrow (in which case we need the object-dtype version anyway since we also can't require numpy 2.0). I would like the proposal to focus on those aspects.

the experimental `pd.NA` sentinel for missing values (which was also [introduced
in pandas 1.0](https://pandas.pydata.org/docs/whatsnew/v1.0.0.html#experimental-na-scalar-to-denote-missing-values)).
However, up to this date, pandas has not yet made the step to use `pd.NA` by
default.

In 2023, [PDEP-10](https://pandas.pydata.org/pdeps/0010-required-pyarrow-dependency.html)
proposed to start using a PyArrow-backed string dtype by default in pandas 3.0
(i.e. infer this type for string data instead of object dtype). To ensure we
could use the variant of `StringDtype` backed by PyArrow instead of Python
objects (for better performance), it proposed to make `pyarrow` a new required
runtime dependency of pandas.

In the meantime, NumPy has also been working on a native variable-width string
data type, which will be available [starting with NumPy
2.0](https://numpy.org/devdocs/release/2.0.0-notes.html#stringdtype-has-been-added-to-numpy).
This can provide a potential alternative to PyArrow for implementing a string
data type in pandas that is not backed by Python objects.

After acceptance of PDEP-10, two aspects of the proposal have been under
reconsideration:

> - Based on user feedback, it has been considered to relax the new `pyarrow`

Reviewer: Is it worth mentioning why this has been objected to? As far as I am aware virtually all objections are due to the installation size effect, and not performance or compatibility.

Author: I can certainly mention _something_, but would prefer to keep that brief to focus here on the strings context and not trigger discussion here about the merits of those objections. 
(for example, it's not only installation _size_, but also the difficulty to install from source in case there are no wheels)

Author: Added "(mostly around installation complexity and size)"

# PDEP-XX: Dedicated string data type for pandas 3.0

- Created: May 3, 2024
- Status: Under discussion

> - Discussion:

Reviewer 2: I see no reason not to use #57073 as the discussion issue as any further discussion will be here and #57073 can now focus on whether to reject PDEP-10 and what to do about the planned improvements to other dtypes.

My assumption is that approval of this PDEP should not, in itself, be a justification to overturn the PDEP-10 decision even though they are very much related and the implementation of the fallback option is only applicable if PDEP-10 is formally rejected.

However, up to this date, pandas has not yet taken the step to use `pd.NA` by
default, and thus the `StringDtype` deviates in missing value behaviour compared
to the default data types.

In 2023, [PDEP-10](https://pandas.pydata.org/pdeps/0010-required-pyarrow-dependency.html)
proposed to start using a PyArrow-backed string dtype by default in pandas 3.0
(i.e. infer this type for string data instead of object dtype). To ensure we
could use the variant of `StringDtype` backed by PyArrow instead of Python
objects (for better performance), it proposed to make `pyarrow` a new required
runtime dependency of pandas.

In the meantime, NumPy has also been working on a native variable-width string
data type, which will be available [starting with NumPy
2.0](https://numpy.org/devdocs/release/2.0.0-notes.html#stringdtype-has-been-added-to-numpy).
This can provide a potential alternative to PyArrow for implementing a string
data type in pandas that is not backed by Python objects.

After acceptance of PDEP-10, two aspects of the proposal have been under
reconsideration:

- Based on user feedback, it has been considered to relax the new `pyarrow`

>   requirement to not be a _hard_ runtime dependency. In addition, NumPy 2.0 can

Author: > I don't think NumPy 2.0 will reduce the need to make pyarrow a dependency for strings

I think it _can_ do that if your motivation for wanting pyarrow is the better performance compared to object-dtype. In that case, numpy 2.0's StringDType can give you a part of the speedup, without requiring pyarrow. 
The discussion in https://github.com/pandas-dev/pandas/issues/57073 also started from that point of view, mentioning numpy 2.0 as an alternative to requiring pyarrow, so based on that my feeling is that what I wrote here is correct (or at least seen as such by some people).

But you are completely right that there are a lot of things that would need to be implemented to make it fully usable for us. That's also the reason that this PDEP does not say to use numpy 2.0, but defers that as a possible future enhancement, to discuss later. And you are also right that it has drawbacks compared to a Arrow based solution (using Arrow memory layout, but not necessary using pyarrow the package), another reason for me personally to again defer that to a separate discussion. 

I just wanted to mention it for the complete context of the string dtype history and discussion. Now, I already mention its existence in the previous paragraph, so could keep it shorter here. 
(and if you have any concrete suggestions to word this better, I am all ears!)

  potentially reduce the need to make PyArrow a required dependency specifically
  for a dedicated pandas string dtype.
- The PDEP did not consider the usage of the experimental `pd.NA` as a
  consequence of adopting one of the existing implementations of the
  `StringDtype`.

For the second aspect, another variant of the `StringDtype` was
[introduced in pandas 2.1](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#whatsnew-210-enhancements-infer-strings)
that is still backed by PyArrow but follows the default missing values semantics
pandas uses for all other default data types (and using `NaN` as the missing

> value sentinel) ([GH-54792](https://github.com/pandas-dev/pandas/issues/54792)).

Reviewer 3: The ``pyarrow_numpy`` StringArray also returns numpy arrays as results for some operations.

I think this is also important to mention.

Author: At this point, I haven't yet mentioned that the original StringDtype returns masked arrays from operations (only that it uses `pd.NA`). I only mention that when going more in detail on this topic in the "Missing value semantics" subsection. Given that, I would also leave it here to the generic "missing value semantics" for the new variant as well (to not make the background section even longer. I can certainly expand the "Missing value semantics"  section if needed)

At the time, the `storage` option for this new variant was called
`"pyarrow_numpy"` to disambiguate from the existing `"pyarrow"` option using `pd.NA`.

This last dtype variant is what you currently (pandas 2.2) get for string data
when enabling the ``future.infer_string`` option (to enable the behaviour which
is intended to become the default in pandas 3.0).

## Proposal

To be able to move forward with a string data type in pandas 3.0, this PDEP proposes:

> 1. For pandas 3.0, we enable a "string" dtype by default, which will use PyArrow

Reviewer 3: Maybe we should mention that ``StringDtype`` is now not experimental?

It's been listed in the docs as experimental for a while now.

Author: > Maybe we should mention that `StringDtype` is now not experimental?

Yeah, Will asked something similar where I answered that this indeed follows from the fact we make it the default (https://github.com/pandas-dev/pandas/pull/58551/files#r1589570042), but I can certainly make that more explicit. 

One nuance, though, is that I personally would only consider the _default variants_ as no longer experimental, but the variants using `pd.NA` can still be kept as experimental (or at least, whether to declare those variants as no longer experimental or not is a bit out of scope for this PDEP)

Reviewer 2: > but the variants using `pd.NA` can still be kept as experimental

+1

Author: Added "Those string dtypes enabled by default will then no longer be considered as
experimental." just below the bullet points

   if installed, and otherwise falls back to an in-house functionally-equivalent
   (but slower) version.
2. This default "string" dtype will follow the same behaviour for missing values
   as our other default data types, and use `NaN` as the missing value sentinel.
3. The version that is not backed by PyArrow can reuse the existing numpy

>    object-dtype backed StringArray for its implementation.

Author: The word "reuse" is certainly a bit ambiguous ;), but given that it's only a small subclass and actually literally uses the same code for the most part of the implementation, I tried to word it like that (I can maybe make it more explicit that it is a shallow subclass, and not a modification of the existing dtype)

>    object-dtype backed StringArray for its implementation.

Reviewer 2: I made another comment, with a simple suggestion, regarding this which you may be happy with instead.

>    object-dtype backed StringArray for its implementation.

Author: I changed your suggestion slightly (because we are in practice mostly adding some code instead of modifying the existing StringArray implementation)

4. We update installation guidelines to clearly encourage users to install

>    pyarrow for the default user experience.

Author: > and do we consider adding a performance warning to the fallback also?

I personally wouldn't do that always / for each method, because that would be super noisy (and in some cases, like smallish data, it doesn't matter that much, so getting those warnings would be annoying).

If we wanted to warn users to gently push them towards installing pyarrow, I think we could do a warning but only 1) raise it once, and 2) only when doing one of the string operations on a big enough dataset (with some threshold).

Now, your question reminds me that the current pyarrow-backed string dtype has those fallback warnings for very specific cases, which I personally think we should stop doing when it becomes the default dtype. Given this is already for the existing implementation (and to keep the many discussion lines here a bit more limited), I opened a separate issue for this: https://github.com/pandas-dev/pandas/issues/58581. 
(but if there is agreement on that other issue, can of course briefly mention that here later)

>    pyarrow for the default user experience.

Reviewer 2: fair point. from the recent user feedback of adding the deprecation warning for the PyArrow requirement, then maybe not having any warnings is wise.

> that the current pyarrow-backed string dtype has those fallback warnings for very specific cases, which I personally think we should stop doing when it becomes the default dtype.

+1

### Default inference of a string dtype

By default, pandas will infer this new string dtype for string data (when
creating pandas objects, such as in constructors or IO functions).

The existing `future.infer_string` option can be used to opt-in to the future
default behaviour:

```python
>>> pd.options.future.infer_string = True
>>> pd.Series(["a", "b", None])
0      a
1      b
2    NaN
dtype: string
```

This option will be expanded to also work when PyArrow is not installed.

### Missing value semantics

Given that all other default data types use NaN semantics for missing values,
this proposal says that a new default string dtype should still use the same
default semantics. Further, it should result in default data types when doing
operations on the string column that result in a boolean or numeric data type
(e.g., methods like `.str.startswith(..)` or `.str.len(..)`, or comparison
operators like `==`, should result in default `int64` and `bool` data types).

Because the original `StringDtype` implementations already use `pd.NA` and
return masked integer and boolean arrays in operations, a new variant of the

> existing dtypes that uses `NaN` and default data types is needed.

Reviewer 4: is -> was to indicate that the referenced variant already exists?

### Object-dtype "fallback" implementation

To avoid a hard dependency on PyArrow for pandas 3.0, this PDEP proposes to keep
a "fallback" option in case PyArrow is not installed. The original `StringDtype`

> backed by a numpy object-dtype array of Python strings can be used for this, and

Reviewer 3: It would be nice to clarify that this is a separate dtype from the original ``string[python]`` dtype, just to make it clear that the original StringDtype is not changing (and still will return masked arrays, and use pd.NA as its missing sentinel)

Author: I tried to clarify in the test that it is indeed a new variant of the string dtype and uses a subclass to reuse most code

only need minor updates to follow the above-mentioned missing value semantics
([GH-58451](https://github.com/pandas-dev/pandas/pull/58451)).

> For pandas 3.0, this is the most realistic option given this implementation is

Reviewer 4: "is already available" -> "has already been available"

already available for a long time. Beyond 3.0, we can still explore further

> improvements such as using nanoarrow or NumPy 2.0, but at that point that is an

Reviewer 3: I would drop the section about nanoarrow (given it is not explained/introduced in the paragraphs beforehand).

If you want to add an explanation, that's also fine with me.

Author: I added a link to the discussion issues for both numpy 2.0 and nanoarrow, so people can find more explanation there if they want.

implementation detail that should not have a direct impact on users (except for
performance).

### Naming

Given the long history of this topic, the naming of the dtypes is a difficult
topic.

In the first place, we need to acknowledge that most users should not need to

> use storage-specific options. Users are expected to specify `pd.StringDtype()`

Reviewer 5: So we are reusing `pd.StringDtype()` in this case right? Is that going to break existing use cases where users have relied on that using pd.NA as a sentinel?

Reviewer 5: Ah thanks - sorry for overlooking that. So I think it goes without saying then that if we go this route we no longer will declare `pd.StringDtype()` experimental? Or are we still trying to keep that reservation knowing even this is not considered a long term design decision?

Author: > So I think it goes without saying then that if we go this route we no longer will declare pd.StringDtype() experimental? 

Yep, given the proposal is to enable this by default, I think that is indeed saying to remove the experimental label (I can mention that somewhere explicitly if that helps)

> Or are we still trying to keep that reservation knowing even this is not considered a long term design decision?

Once we have _a_ `"string"`, we will always have one, I think. That aspect is the long term decision this PDEP is proposing. We might change later the missing value semantics, but that doesn't mean the string dtype proposed here is still experimental (just like our default "int64" dtype is not experimental). At the time that we would decide to enable new missing value semantics by default, then `"string"` will "simply" start meaning something differently.

or `"string"`, and that will give them their default string dtype (which
depends on whether PyArrow is installed or not).

But for testing purposes and advanced use cases that want control over this, we
need some way to specify this and distinguish them from the other string dtypes.
Currently, the `StringDtype(storage="pyarrow_numpy")` is used, where
"pyarrow_numpy" is a rather confusing option.

> TODO see if we can come up with a better naming scheme

Reviewer 4: `StringDtype(storage="pyarrow", semantics="numpy")`?  or instead of semantics, could use "na_value=np.nan`

Reviewer 2: If i'm understanding correctly about the motivation for the change in dtype (improved overall user experience), then moving forward I suspect that when we can have improved/native dtypes for other data types (nested, date, etc) that the same logic would need to apply, i.e. we would need to have a variants of these with NumPy semantics.

Now this probably falls under PDEP-13 but if we have `semantics` as a argument (that users would see and use) we could still end up with columns using different missing value indicators?

Reviewer 2: > `StringDtype(storage="pyarrow", semantics="numpy")`? or instead of semantics, could use "na_value=np.nan`

or maybe "nullable=[True|False]"

However, at the moment, we distinguish the nullable data types for the other dtypes (int, float, etc) with capitalization and so for consistency could also consider string/String as the dtypes.

Author: > `StringDtype(storage="pyarrow", semantics="numpy")`? or instead of semantics, could use "na_value=np.nan`

@jbrockmendel good point that we can also use other keywords than just `storage` to make the distinction

> if we have `semantics` as a argument (that users would see and use) we could still end up with columns using different missing value indicators?

Only if users explicitly specify a non-default value for this, and never by default. This is the same with whatever option we come up with (eg also when using `dtype_backend="pyarrow"` or explicitly asking for one of the masked dtypes with `dtype=Int64` or .. you can end up with a DataFrame with columns with mixed semantics)

> we distinguish the nullable data types for the other dtypes (int, float, etc) with capitalization and so for consistency could also consider string/String as the dtypes.

Yeah, only unfortunately to be consistent with the other dtypes where we use capitalization, it would need to be `"string"` for the new NaN-based dtype, and `"String"` for the "nullable" NA-based variant. And so that doesn't help with backwards compatibility, because `"string"` right now means the nullable dtype. Given that, I would personally not use capitalization here (which also only is a solution for the string alias naming, not for the `StringDtype(..)` API)

---

To keep the sub-discussions manageable, I moved this specific topic out of this inline comment thread, and into it's own issue: https://github.com/pandas-dev/pandas/issues/58613

## Alternatives

### Why not delay introducing a default string dtype?

To avoid introducing a new string dtype while other discussions and changes are
in flux (eventually making pyarrow a required dependency? adopting `pd.NA` as
the default missing value sentinel? using the new NumPy 2.0 capabilities?), we
could also delay introducing a default string dtype until there is more clarity
in those other discussions.

However:

1. Delaying has a cost: it further postpones introducing a dedicated string
   dtype that has massive benefits for our users, both in usability as (for the

>    significant part of the user base that has PyArrow installed) in performance.

Reviewer 3: I don't think we can just claim this. I don't disagree, but this should be backed up more.

At least from the feedback received from #57073 and the other issue, there's at least a significant part of the user base that doesn't use strings.

There's also a significant chunk of the population that can't install pyarrow (due to size requirements or exotic platforms or whatever).

Reviewer 5: I am not sure this argument is that convincing either, although for slightly different reasons. I don't think we need to feel rushed for the next release

Author: > I don't think we can just claim this. I don't disagree, but this should be backed up more.

@lithomas1 can you clarify which part of the paragraph you think requires more backing up? 
The fact that I say a "significant" part of our user base has pyarrow installed?

I don't think we can ever know exact numbers for this, but one data point is that pandas currently has 210M monthly downloads and pyarrow has 120M monthly downloads. Of course not all of those pyarrow users are also using pandas, but let's just assume that half of those pyarrow downloads come from people using pandas, that would mean that around 30% for our users already have pyarrow installed, which I would consider as a "significant part". 
(and my guess is that for people working with larger datasets, where the speed of pyarrow becomes more important, this percentage will be higher, for example because of using the parquet IO)

But anyway, we are never going to know this exact number, but IMO we _do_ know that a significant part of our userbase has pyarrow and will benefit from using that by default.

> there's at least a significant part of the user base that doesn't use strings.

Yes, and then this PDEP is not relevant for them. But it's not because some users don't use strings, that we shouldn't improve the life of those users that _do_ use strings (so just not really understanding how this is a relevant argument)

> There's also a significant chunk of the population that can't install pyarrow

Yes, and this PDEP addresses that by allowing a fallback when pyarrow is not installed.

> I am not sure this argument is that convincing either, although for slightly different reasons.

@WillAyd can you then clarify which other reasons?

Reviewer 5: My other reason is that I don't think there is ever a rush to get a release out; we have historically never operated that way

Reviewer 3: > > I don't think we can just claim this. I don't disagree, but this should be backed up more.
> 
> @lithomas1 can you clarify which part of the paragraph you think requires more backing up? The fact that I say a "significant" part of our user base has pyarrow installed?

It'd be nice to add how much perf benefits Arrow strings are expected to bring (e.g. 20%? 2x? 10x?). 
Putting in the part about how many users have pyarrow would also help.

It'd also be good to elaborate on the usability part. IIUC, the main benefit here is not having to manually check element to see whether your object dtype'd column contains strings (since I think all the string methods work on object dtype'd columns).

I think it's also fair to amend this part to say "massive benefits to users that use strings" (instead of in general).

Reviewer 5: Benchmarks are going to be highly dependent on usage and context. If working in an Arrow native ecosystem, the speedup of strings may be a factor over 100x. If working in a space where you have to copy back and forth a lot with NumPy, that number goes way down.  

I think trying to set expectations on one number / benchmark for performance is futile, but generally Arrow only helps, and makes it so that we as developers don't need to write custom I/O solutions (eg: ADBC Drivers, parquet, read_csv with pyarrow all work with Arrow natively with  no extra pandas dev effort)

Author: >> It'd be nice to add how much perf benefits Arrow strings are expected to bring (e.g. 20%? 2x? 10x?).
>
> Benchmarks are going to be highly dependent on usage and context.

Indeed, for single operations you can easily get a >10x speedup, but of course a typical workflow does not consist of just string operations, and the overall speedup depends a lot (see this slide for one small example comparison (https://phofl.github.io/pydata-berlin/pydata-berlin-2023/intro.html#74) and this blogpost from Patrick showing the benefit in a dask example workflow (https://towardsdatascience.com/utilizing-pyarrow-to-improve-pandas-and-dask-workflows-2891d3d96d2b).

> but generally Arrow only helps, and makes it so that we as developers don't need to write custom I/O solutions

That is often true, but _except_ for strings ;). 
For strings, the faster compute kernels will still give a lot of value even if your IO wasn't done through Arrow (and give a lot more value compared to using pyarrow for numeric data)

>    dtype that has massive benefits for our users, both in usability as (for the
>    significant part of the user base that has PyArrow installed) in performance.

Reviewer 2: ```suggestion
   dtype that has massive benefits for our users, both in usability and, for users that already have PyArrow installed or have no issues installing PyArrow, in performance.
```

2. In case we eventually transition to use `pd.NA` as the default missing value
   sentinel, we will need a migration path for _all_ our data types, and thus
   the challenges around this will not be unique to the string dtype and

>    therefore not a reason to delay this.

Reviewer 2: other than maybe this is the only default data type that will be using PyArrow (I mention elsewhere that maybe we need to discuss the consequences of this)

Author: > other than maybe this is the only default data type that will be using PyArrow

Which will actually make it _easier_ to transition to use `pd.NA`, since we already have a variant of the dtype that does that?

Reviewer 2: Yeah. This text is fine as is.

`object` dtype (such as `ser.dtype == object`) will need to be updated.

To allow testing your code in advance, the
`pd.options.future.infer_string = True` option is available.

Otherwise, the actual string-specific functionality (such as the `.str` accessor
methods) should all keep working as is. By preserving the current missing value
semantics, this proposal is also backwards compatible on this aspect.

One other backwards incompatible change is present for early adopters of the
existing `StringDtype`. In pandas 3.0, calling `pd.StringDtype()` will start
returning the new default string dtype, while up to now this returned the
experimental string dtype using `pd.NA` introduced in pandas 1.0. Those users
will need to start specifying a keyword in the dtype constructor if they want to
keep using `pd.NA` (but if they just want to have a dedicated string dtype, they
don't need to change their code).

## Timeline

The future PyArrow-backed string dtype was already made available behind a feature
flag in pandas 2.1 (by `pd.options.future.infer_string = True`).

Some small enhancements or fixes (or naming changes) might still be needed and

> can be backported to pandas 2.2.x.

Reviewer 3: This part of the plan worries me a little.

Maybe it would be better to cut off a 2.3 from 2.2.x.

I think there's a significant proportion of the downloads for 2.2 that aren't on the latest patch release.
I think there's ~ 1/3 of the downloads that are fetching 2.2.0.

Reviewer 3: Also,
it would be good to mention which version of pandas is expected to have  ``infer_string`` be able to infer to the fallback option.

Reviewer 2: a 2.3 release (maybe around the same time as 3.0rc) sounds reasonable. 

If the features/bugfixes added to 2.3 are limited to the string dtype then we shouldn't need many patch releases. We may not need to fix any string dtype related issues that are fixed for 3.0 as these will be behind a flag in 2.3 and so shouldn't break existing code.

On the other hand, as these features are behind a flag, maybe releasing a 2.3 would not gain the field testing we hope for.

And therefore, instead of doing a 2.3, planning for at least a couple of release candidates for 3.0 would better achieve this.

Reviewer 3: @jorisvandenbossche 

Thoughts on this?

Author: > Maybe it would be better to cut off a 2.3 from 2.2.x.

Yes, if we still plan to add a deprecation warning and change the naming scheme in `StringDtype`, calling that 2.3.0 sounds as the best option (I had been planning to propose doing a 2.3.0 (from the 2.2.x branch) anyway to bump the warning for CoW from DeprecationWarning to FutureWarning)

> # PDEP-XX: Dedicated string data type for pandas 3.0

Reviewer 6: ```suggestion
# PDEP-14: Dedicated string data type for pandas 3.0
```

  dedicated pandas string dtype.
- The PDEP did not consider the usage of the experimental `pd.NA` as a
  consequence of adopting one of the existing implementations of the
  `StringDtype`.

For the second aspect, another variant of the `StringDtype` was
[introduced in pandas 2.1](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#whatsnew-210-enhancements-infer-strings)
that is still backed by PyArrow but follows the default missing values semantics
pandas uses for all other default data types (and using `NaN` as the missing
value sentinel) ([GH-54792](https://github.com/pandas-dev/pandas/issues/54792)).
At the time, the `storage` option for this new variant was called
`"pyarrow_numpy"` to disambiguate from the existing `"pyarrow"` option using `pd.NA`.

This last dtype variant is what you currently (pandas 2.2) get for string data
when enabling the ``future.infer_string`` option (to enable the behaviour which
is intended to become the default in pandas 3.0).

## Proposal

To be able to move forward with a string data type in pandas 3.0, this PDEP proposes:

1. For pandas 3.0, we enable a "string" dtype by default, which will use PyArrow

>    if installed, and otherwise falls back to an in-house functionally-equivalent

Reviewer 6: If PyArrow is installed but the user calls a string operation that isn't supported via PyArrow (e.g. functionality via a keyword argument specified), will the operation raise or continue to fall back to the "in-house functionally-equivalent" (and re-return a PyArrow string type)?

Author: Do we actually have such a case right now where it falls back because of pyarrow not supporting a certain string kernel and the result is a new string array? Looking at the ArrowStringArray implementation, I only see the `fallback_performancewarning` being used in replace and contains, so both predicates returning bools. 
But anyway, the answer would be to indeed re-return the pyarrow-backed string dtype, I would say

Reviewer 6: I also see (where we missed adding `fallback_performancewarning`)

* `repeat` where `repeats=` is a sequence of ints
* `slice` where `stop=None`
* `removeprefix` for pyarrow < 13
* `find` for some `start`/`end` combination

But sure, I guess in these cases it would be OK to go from pyarrow -> object -> pyarrow to support these ops

>    (but slower) version.

Reviewer 7: Isn't the "in-house functionally-equivalent (but slower) version" the current implementation based on numpy 1.x in version 2.2, but we now make the dtype `string` instead of `object` ?

Author: In the section below that expands on this object-type based implementation ("Object-dtype "fallback" implementation"), there is a bit longer explanation and I also link to the open PR implementing this: https://github.com/pandas-dev/pandas/pull/58451

(and fwiw, this is not specific to numpy 1.x, it will also work on numpy 2.x, it just uses `object` dtype)

Reviewer 7: I have to admit that I am pretty confused by these various implementations and the corresponding semantics used to describe them.

I'd suggest having a summary of what exists today in pandas 2.2 (strings with object dtype, `np.nan`), the current `StringDtype` with `pd.NA`, the "experimental" `pyarrow` based implementations (with both `pd.NA` and `np.nan` being available), and anything else, and what is proposed would be available due to this PDEP, and how it might change in the future due to however we decide to handle missing values in the future, as well as `nanoarrow` and `numpy 2.0` strings.

Author: Current situation (pandas 2.2) is:

* Current default: `object` dtype with np.nan or None
* Experimental opt-in string dtypes using `pd.NA`: `StringDtype()` with storage being `"python"` (default, object dtype under the hood) or `"pyarrow"`
* Future string dtypes using `NaN` (behind `pd.options.future.infer_string = True`): `StringDtype()` with storage being `"pyarrow_numpy"`
  * This is the dtype that is essentially being proposed in this PDEP, but already exists since pandas 2.1 ([#54792](https://github.com/pandas-dev/pandas/issues/54792))

And then this PDEP also describes to add an extra option for the third bullet point, i.e. having a `StringDtype()` using NaN but backed by object-dtype array instead of pyarrow, which I dubbed (for now) `"python_numpy"` in the PR adding this (https://github.com/pandas-dev/pandas/pull/58451), but that name is still being discussed.

As a starter (still thinking about how to make this clearer in the PDEP), does the above clarify it for you?

Note that the above listing leaves out `pd.ArrowDtype(<some pyarrow string type>)`, because that is not really relevant for the discussion right now and I am not sure mentioning it will help, but that is yet another way to store strings.

Reviewer 7: > As a starter (still thinking about how to make this clearer in the PDEP), does the above clarify it for you?
> 
Yes, although I'd suggest for the PDEP a table that indicates the constructor version of the dtype (e.g., `dtype=object`, or `dtype=pd.StringDtype()`), and the string representation of the dtype.

> Note that the above listing leaves out `pd.ArrowDtype(<some pyarrow string type>)`, because that is not really relevant for the discussion right now and I am not sure mentioning it will help, but that is yet another way to store strings.

Is that equivalent to having a pure `"pyarrow"` backed string that uses `pd.NA` as the null semantics?  If not, where does that fit it?  Or is it not available?

Author: > Is that equivalent to having a pure `"pyarrow"` backed string that uses `pd.NA` as the null semantics? If not, where does that fit it? Or is it not available?

`StringDtype(storage="pyarrow")` is also a "pure pyarrow backed string that uses NA", to be clear. But so 
`StringDtype(storage="pyarrow")`and `ArrowDtype(pa.(large_)string())` are essentially equivalent, except for using a different dtype and array class.

use storage-specific options. Users are expected to specify `pd.StringDtype()`
or `"string"`, and that will give them their default string dtype (which
depends on whether PyArrow is installed or not).

But for testing purposes and advanced use cases that want control over this, we
need some way to specify this and distinguish them from the other string dtypes.
Currently, the `StringDtype(storage="pyarrow_numpy")` is used, where
"pyarrow_numpy" is a rather confusing option.

TODO see if we can come up with a better naming scheme

## Alternatives

### Why not delay introducing a default string dtype?

To avoid introducing a new string dtype while other discussions and changes are
in flux (eventually making pyarrow a required dependency? adopting `pd.NA` as
the default missing value sentinel? using the new NumPy 2.0 capabilities?), we
could also delay introducing a default string dtype until there is more clarity
in those other discussions.

However:

1. Delaying has a cost: it further postpones introducing a dedicated string
   dtype that has massive benefits for our users, both in usability as (for the
   significant part of the user base that has PyArrow installed) in performance.

> 2. In case we eventually transition to use `pd.NA` as the default missing value

Reviewer 5: > the challenges around this will not be unique to the string dtype and
   therefore not a reason to delay this.

I might be missing the intent but I don't understand why the larger issue of NA handling means we should be faster to implement this

Author: > I don't understand why the larger issue of NA handling means we should be faster to implement this

It's not a reason to do it "faster", but I meant to say that the discussion regarding NA is not a reason to do it "slower" (to delay introducing a dedicated string dtype)

Reviewer 5: I think the flip side is that if we aren't careful about the NA handling we can introduce some new keywords / terminology that makes it very confusing in the long run (which is essentially one of the problems with our strings naming conventions)

As a practical example, if we decided we wanted `semantics=` as a keyword argument to `StringDtype` in this PDEP to move the NA discussion along, that might be counter-productive when we look at more data types and decide `semantics=` was not a clear way to allow datetime data types to support `pd.NaT` as the missing value. 

(not saying the above is necessarily the truth, just cherry picking from conversation so far)

   sentinel, we will need a migration path for _all_ our data types, and thus
   the challenges around this will not be unique to the string dtype and
   therefore not a reason to delay this.

### Why not use the existing StringDtype with `pd.NA`?

Wouldn't adding even more variants of the string dtype will make things only more
confusing? Indeed, this proposal unfortunately introduces more variants of the

> string dtype. However, the reason for this is to ensure the actual default user

Reviewer 5: This just retroactively clarifies the reasoning for `string[pyarrow_numpy]` to have existed in the first place right? Or is it supposed to be hinting at some other feature that the implementation details of the PDEP is proposing?

Author: Yes, it's indeed explaining why we did this, which is of course "retroactively" given I was asked to write this PDEP partly for changes that have already been released. So a big part of the PDEP is retroactively in that sense (which it not necessarily helping to write it clearly ..).

Reviewer 2: > Or is it supposed to be hinting at some other feature that the implementation details of the PDEP is proposing?

however, more importantly, the PDEP makes this (the already added dtype) the default in 3.0. It would remain behind the future flag for the next release if enough people feel we are not ready.

bool, etc dtypes). This would lead to a very confusing default experience.

With the proposed new variant of the StringDtype, this will ensure that for the
_default_ experience, a user will only see only 1 kind of integer dtype, only
kind of 1 bool dtype, etc. For now, a user should only get columns with an
`ArrowDtype` and/or using `pd.NA` when explicitly opting into this.

## Backward compatibility

The most visible backwards incompatible change will be that columns with string
data will no longer have an `object` dtype. Therefore, code that assumes
`object` dtype (such as `ser.dtype == object`) will need to be updated.

To allow testing your code in advance, the
`pd.options.future.infer_string = True` option is available.

Otherwise, the actual string-specific functionality (such as the `.str` accessor
methods) should all keep working as is. By preserving the current missing value
semantics, this proposal is also backwards compatible on this aspect.

One other backwards incompatible change is present for early adopters of the
existing `StringDtype`. In pandas 3.0, calling `pd.StringDtype()` will start
returning the new default string dtype, while up to now this returned the

> experimental string dtype using `pd.NA` introduced in pandas 1.0. Those users

Reviewer 5: Historically you would get this by using `dtype="string"` too right? I'm a little wary that we are underestimating the scope of how breaking this could be; I didn't even realize we considered that dtype experimental all this time

Author: A deprecation warning about what exactly?

Reviewer 5: > at the expense of increasing the scope of breaking changes for the smaller subset of users that was already using `dtype="string"`

This is where I am a little uncomfortable - I don't know how to measure the size of that, but I am wary of assuming it is not a signifcant number of users. The fact that "string" returns NA as a missing value is a documented difference in our code base:

https://pandas.pydata.org/docs/dev/user_guide/text.html#behavior-differences

And its usage has been promoted for quite some time:

https://stackoverflow.com/a/60553529/621736
https://towardsdatascience.com/why-we-need-to-use-pandas-new-string-dtype-instead-of-object-for-textual-data-6fd419842e24
https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.1.0.html#all-dtypes-can-now-be-converted-to-stringdtype

> If we don't want to make `dtype="string"` breaking, then either we need to come up with a different name for the dtype (not using "string", like "utf8" or "text"), or either we need to delay introducing a default string dtype until after we have agreement on the NA discussions.

Yea none of these options are great...but out of them I still would probably prefer waiting. I think right now we are marching down a path of "string" missing values:

  1. Returning pd.NA today
  2. Returning np.nan with this PDEP (granted those changes are already in main)
  3. Going back to returning pd.NA with the NA PDEP

Reviewer 5: Adding on that this is all done with zero deprecation; using dtype="string" for 4 years has allowed users to intentionally work around the shortcomings of the NumPy type system. As far as I understand this PDEP would mean that:

```python
>>> ser = pd.Series(["foo", "bar", None], dtype="string")
>>> ser.str.isalnum()
0    True
1    True
2    <NA>
dtype: boolean
```

is going to revert back to:

```python
>>> ser.str.isalnum()
0    True
1    True
2    None
dtype: object
```

with no warning or heads up to users who have intentionally been using the "string" data type all this time. 

Even if we agreed as a team that we were OK with going back to the NumPy type system (I don't want to...but speaking hypothetically) not doing any type of deprecation at all just to hit an arbitrary release date is a showstopper for me

Author: > is going to revert back to:

My intention is that it would return the following (see https://github.com/pandas-dev/pandas/issues/54805, as one of the design issues listed in https://github.com/pandas-dev/pandas/issues/54792 that we still needed to resolve):
```python
>>> ser.str.isalnum()
0    True
1    True
2    False
dtype: bool
```

which is still different (i.e. not having a missing value in the result), but at least it gives a boolean result (and is consistent with how NaN propagation works in other predicates for our default dtypes). 
Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.

Author: While there will definitely be code breakages, I think that pattern will be be the case for many typical workflows using pandas methods. Whether the string column uses NaN or NA, many workflows will continue to work fine because our methods are designed to work with them (eg `isna` will work either way whether the column uses NaN or NA). There are of course specific behaviours that will change, like if you check for the exact scalar for a missing value, or the propagation in comparisons. But unless your code has a specific check for `pd.NA` in it, it is likely going to keep working, I think.

That is not to say that those breakages are not real. But my tradeoff that I am currently making is that I think this is worth it for the benefit of having a "string" dtype in 3.0 for _all_ users at the cost of some breakage for people using the experimental dtype (and note, they will always be able to update their code to continue to use those NA-based dtypes, by slightly changing the way the dtype is specified, but then without having to change anything else in their code).

In the end:

> Yea none of these options are great...but out of them I still would probably prefer waiting.

is of course a completely fine stance, and it's OK to disagree on that trade-off (and my PDEP is proposing to not wait).

Reviewer 5: > Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.

I am confused - this works today if you use `dtype="string"` but this PDEP is going to break that (?).

We introduced dtype="string" 4 years ago intentionally with pd.NA to avoid these nuances of the NumPy type system. For users that have been current with their use of pandas, this part of the PDEP especially is not true:

> Otherwise, the actual string-specific functionality (such as the `.str` accessor
methods) should all keep working as is. By preserving the current missing value
semantics, this proposal is also backwards compatible on this aspect.

since for those users in particular it is taking a step back...

I also disagree with calling the existing dtype experimental - it has been around for 4 years and served a purpose. The fact that it says experimental in the documentation is just an oversight; it's made its way through a few major releases so far without any material changes

Reviewer 5: I'm also cognizant that this PDEP is not exclusively about missing handling, but the unfortunate fact today is that without PDEP-13 and PDEP-XX coming before this then our data types and na markers are bundled together.

With that being the case, if the intent here is to step back from pd.NA -> np.nan but then later return to pd.NA as a default I think it should be explicitly stated.

Reviewer 7: > We introduced dtype="string" 4 years ago intentionally with pd.NA to avoid these nuances of the NumPy type system. For users that have been current with their use of pandas, this part of the PDEP especially is not true:
> 
> > Otherwise, the actual string-specific functionality (such as the `.str` accessor
> > methods) should all keep working as is. By preserving the current missing value
> > semantics, this proposal is also backwards compatible on this aspect.
> 
> For those users in this is taking a step back...

I think that @WillAyd has a valid point.

If people were NOT using `dtype="string"`, but rather had all strings as `dtype="object"`, then this PDEP maintains the behavior of how strings are processed in 2.x, but just changes the dtype.

But if people had opted-in to using `dtype="string"`, then this PDEP will change the behavior.

It is most likely that there are more people who handle strings with `dtype="object"` than `dtype="string"`., so the comment above about "actual string-specific functionality...should all keep working as is" is true for that group.

But we have a problem with the people who use `dtype="string"` today.

Reviewer 4: I would be fine with changing the behavior for `dtype="string"` and `dtype=pd.StringDtype()`, but would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Author: > > Having it as a bool will ensure that typical operations (eg `ser[ser.str.isalnum()]`) will continue to work as expected.
> 
> I am confused - this works today if you use `dtype="string"` but this PDEP is going to break that (?).

No, my answer to your example snippet was trying to explain how I would ensure this does _not_ break (if we return `bool` column instead of object dtype with True/False/NaN will ensure that filtering keeps working).

> [@jbrockmendel] would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Patrick actually already changed `dtype="string"` to mean the future default dtype as well (so not only the case of inference without specified dtype). 
(whether it is already in main or not is not too important I think, if we feel it is better to change this. But personally I think `dtype="string"` meaning something different than the default string dtype you get without specifying the dtype is going to be very confusing ..)

Reviewer 7: > I would be fine with changing the behavior for `dtype="string"` and `dtype=pd.StringDtype()`, but would _also_ be OK with only changing it for users who don't specify a dtype, i.e. `pd.Series(['a', 'b', None])`, which IIUC is the only part that is actually in main.

Just to make sure that I understand what @jbrockmendel is suggesting in the second half of his statement:
- If a user doesn't specify a dtype, but we infer the dtype to be strings, we use the current "pyarrow_numpy" type, which means storage is "pyarrow" and missing values are `np.nan`.  If `pyarrow` is not installed, it uses python strings and missing values are `np.nan`.  Then, for current users who don't specify a dtype, they get the current behavior with respect to missing values.
- If a user DOES specify `dtype="string"` or `pd.StringDtype()`, we use the current behavior, which means we use missing values as `pd.NA()`.  Then storage is "pyarrow" and if `pyarrow` is not installed, it uses python strings.  But either way, missing values are represented by `pd.NA`.

Reviewer 7: > But personally I think `dtype="string"` meaning something different than the default string dtype you get without specifying the dtype is going to be very confusing ..)

I think we have to carefully specify what the user specifies in a `dtype` argument and how that gets interpreted, versus what we return as the `dtype` when they look at `Series.dtype`.  

So we could have a mapping that says

| User specifies `dtype`= | pandas returns `Series.dtype` |
| ------------------------|-------------------------------|
| Unspecified             | `"string[pyarrow_numpy]"` OR `"string[python]"`     |
| `"string"`              | `"string[pyarrow]"`           |
| StringDtype("pyarrow")  | `"string[pyarrow]"`           |
| StringDtype("python")   | `"string[python]"`            |
| StringDtype("pyarrow_numpy") | `"string[pyarrow_numpy]"` |

The first row depends on whether `pyarrow` is installed.
For the second, third and fifth rows, if `pyarrow` is not installed, we raise an Exception.

Separately, we can then debate what the values in the second column should look like in #58613  .  I personally am not a fan of `"pyarrow_numpy"`

Reviewer 5: > No, my answer to your example snippet was trying to explain how I would ensure this does _not_ break (if we return `bool` column instead of object dtype with True/False/NaN will ensure that filtering keeps working).

Ah OK - I didn't realize you were proposing that change be a part of this PDEP, just thought it was an idea you had for the future. But that's a completely new behavior...and then begs the question of do we go back and change dtype=object to have that same behavior or just have dtype="string" exclusively have it. Ultimately we end up with the same issue

Reviewer 3: Yeah, I also agree with Will that it's not fair to change this without warning for people already using "string".
(pd.NA is also a big selling point of the ``dtype="string"`` too)

Maybe a good compromise would be to use ``string[pyarrow]`` under the hood  for those users (if they had it installed)?

If we were to move ahead with the move to nullable dtypes in general, I worry that this changing of the na value for ``dtype="string"`` from pd.NA -> np.nan -> pd.NA will cause a lot of confusion.

If we were to do 2.3 (like I suggested below), this might be addressable there (with a deprecation).

Author: Still adding some deprecation warnings in 2.x for current users of StringDtype is something we certainly could do. I am personally ambivalent about it, but fine with adding it if others think that is better (I do think it might become quite noisy, and it also does not change the fact that 3.0 will switch from NA to NaN)

Author: > I think we have to carefully specify what the user specifies in a `dtype` argument and how that gets interpreted, versus what we return as the `dtype` when they look at `Series.dtype`.
> 
> So we could have a mapping that says

I created a variant of that table https://github.com/pandas-dev/pandas/issues/58613#issuecomment-2100453677 with a concrete proposal

> For the second, third and fifth rows, if `pyarrow` is not installed, we raise an Exception.

(for clarity, this "second" row referred to specifying a dtype with `"string"`) 
If you explicitly ask for pyarrow, then yes raising an exception is fine and expected. But a generic `"string"` (or `StringDtype()`) has to mean "whatever string dtype that is the default" and so cannot raise an exception if pyarrow is not installed, but should return the object-dtype based fallback.

will need to start specifying a keyword in the dtype constructor if they want to
keep using `pd.NA` (but if they just want to have a dedicated string dtype, they
don't need to change their code).

## Timeline

The future PyArrow-backed string dtype was already made available behind a feature
flag in pandas 2.1 (by `pd.options.future.infer_string = True`).

Some small enhancements or fixes (or naming changes) might still be needed and
can be backported to pandas 2.2.x.

The variant using numpy object-dtype could potentially also be backported to

> 2.2.x to allow easier testing.

Reviewer 7: Don't you mean "numpy 2.0 string-dtype" ?  Because the "numpy object-dtype" is currently there?  This labeling is confusing.

Given the discussions elsewhere about where the names are now for the dtype `"string[pyarrow]"`, `"string[pyarrow_numpy]"`, etc. which I can't keep track of, I think that the nomenclature in terms of strings that work should be specified, comparing what is in 2.2 to what would be implemented as a result of this PDEP.

The possible strings are confusing.  Which strings can be used in a `dtype` argument in a constructor or `astype()` ?  Which strings would be seen by users when they do `Series.dtype` ?  There is what currently exists in pandas 2.2, and what would exist based on this PDEP, and I'm not seeing what the conclusion of that would be.
